### PR TITLE
Fixes reward and text gathering in distributed training

### PR DESCRIPTION
**Pull request by edbeeching**
This PR fixes two issues:

- **Rewards** were being wrongly summed instead of gathered. While this did not affect the mean, other stats such as std and the distribution plots in wandb were affected.
- **Texts** were not being gathered across processes, leading to text only being reported from the main process.

I have tested in two settings [single node 8 GPU](https://wandb.ai/edbeeching/trl/runs/5ovy2waw/overview?workspace=user-edbeeching) and [single node 1 GPU](https://wandb.ai/edbeeching/trl/runs/6zz4yyh2?workspace=user-edbeeching). I launched both with `--ppo-config.mini-batch-size=1 --ppo-config.batch-size=2` to confirm the gather is working correctly.

This may impact #847

### CODE REVIEW

File: trl/trainer/ppo_trainer.py
Comment (vwxyzjn) [inline trl/trainer/ppo_trainer.py:1328]: Similar as above.

File: trl/trainer/ppo_trainer.py
Comment (vwxyzjn) [inline trl/trainer/ppo_trainer.py:1383]: What is this doing?

File: trl/trainer/ppo_trainer.py
Comment (vwxyzjn) [inline trl/trainer/ppo_trainer.py:1308]: Maybe all these line of code can be replaced with self.accelerator.gather(rewards)


```suggestion
            # Gather rewards if distributed
            rewards = self.accelerator.gather(rewards).flatten()
```

**Review by vwxyzjn — COMMENTED** [review]
Very nice PR! Thanks @edbeeching this will make logging much better

File: trl/trainer/ppo_trainer.py
Comment (edbeeching) [reply trl/trainer/ppo_trainer.py:1308]: Ah nice, I was not aware accelerate has this helper method, I will clean up.

File: trl/trainer/ppo_trainer.py
Comment (edbeeching) [reply trl/trainer/ppo_trainer.py:1383]: Here I am sending the query & response from process n>1 to process 0. This is two way, perhaps there is a way to do this one way with send / recv? I don't often use those operations.

**Comment by lewtun** [comment]
Begone stale bot 🪄!

**Comment by edbeeching** [comment]
Yes, I will get back to this when I have the time.

**Comment by edbeeching** [comment]
@vwxyzjn I have rebased and updated the PR if you can take another look.

File: trl/trainer/ppo_trainer.py
Comment (vwxyzjn) [inline trl/trainer/ppo_trainer.py:1380]: @lvwerra do these code do anything?

**Review by vwxyzjn — APPROVED** [review]
@edbeeching I have refactored just a bit :) There seems to be some unused code that I will have @lvwerra take a look at but otherwise LGTM.

Thanks a lot for the PR!

File: trl/trainer/ppo_trainer.py
Comment (lvwerra) [reply trl/trainer/ppo_trainer.py:1380]: cc @younesbelkada :)

**Review by lvwerra — APPROVED** [review]
Looks good to me, I'll let @younesbelkada have a look too!

File: trl/trainer/ppo_trainer.py
Comment (lewtun) [inline trl/trainer/ppo_trainer.py:1343]: turbo nit to follow HF conventions to use descriptive names
```suggestion
                    for batch in batch_list:
```

**Review by lewtun — APPROVED** [review]
LGTM with a turbo nit about variable names :)

I also have the same question as @vwxyzjn about the remove of some distributed code, but @younesbelkada is the best person to answer that

**Comment by lvwerra** [comment]
Is this ready to be merged?

**Comment by vwxyzjn** [comment]
Merged! Thanks @edbeeching @lewtun @lvwerra !

**Pull request merged**
